parsons/action_builder/action_builder.py

```diff
@@ -0,0 +1,351 @@
+            raise ValueError("Must provide name or given name if inserting new record")
+
+        campaign = self._campaign_check(campaign)
+
+        url = f"campaigns/{campaign}/people"
+
+        if "person" not in data:
+            # The POST data must live inside of person key
+            data["person"] = {}
+
+        if identifiers:
+            # Updating an existing record
+
+            if isinstance(identifiers, str):
+                # Ensure identifiers is a list, even if it will usually just be one item
+                identifiers = [identifiers]
+
+            # Default to assuming identifier comes from Action Builder and add prefix if missing
+            identifiers = [
+                f"action_builder:{x}" if ":" not in x else x for x in identifiers
+            ]
+            data["person"]["identifiers"] = identifiers
+
+        if entity_type:
+            # Inserting a new record
+            data["person"]["action_builder:entity_type"] = entity_type
+
+        return self.api.post_request(url=url, data=json.dumps(data))
+
+    def add_tags_to_record(
+        self, identifiers, tag_name, tag_field, tag_section, campaign=None
+    ):
+        """
+        Add a tag (i.e. custom field value) to an existing entity record in Action Builder. The
+        tags, along with their category/field and section/group, must already exist (except for
+        date fields). If fewer fields are provided than tags, the excess tags will be applied to
+        the final field provided. Similarly, the final section will be applied to any fields or
+        tags in excess.
+        `Args:`
+            identifiers: list
+                The list of strings of unique identifiers for a record being updated. ID strings
+                will need to begin with the origin system, followed by a colon, e.g.
+                `action_builder:abc123-...`.
+            tag_name: str or list
+                The name(s) of the new tag(s), i.e. the custom field value(s).
+            tag_field: str or list
+                The name(s) of the tag category(ies), i.e. the custom field name(s). Must have
+                equal or fewer items as `tag_name`.
+            tag_section: str or list
+                The name(s) of the tag section(s), i.e. the custom field group name(s). Must have
+                equal or fewer items as `tag_section`.
+            campaign: str
+                Optional. The 36-character "interact ID" of the campaign whose data is to be
+                retrieved or edited. Not necessary if supplied when instantiating the class.
+        `Returns:`
+            Dict containing Action Builder entity data of the entity being tagged.
+        """
+
+        # Ensure all tag args are lists
+        tag_name = tag_name if isinstance(tag_name, list) else [tag_name]
+        tag_field = tag_field if isinstance(tag_field, list) else [tag_field]
+        tag_section = tag_section if isinstance(tag_section, list) else [tag_section]
```

Comment:
what would happen if someone provided data as the wrong type. I'm thinking like a tuple which would then become a list of tuples. If we passed that to the api, would we get really weird tags or would it fail?

Reply:
It really should fail, because the resulting data would essentially be attempting to apply a tag that doesn't exist.